compiler/rustc_middle/src/thir.rs

```diff
@@ -746,6 +747,11 @@ pub enum PatKind<'tcx> {
         value: mir::Const<'tcx>,
     },
 
+    InlineConstant {
+        value: mir::UnevaluatedConst<'tcx>,
+        subpattern: Box<Pat<'tcx>>,
+    },
```

Comment:
What do the fields represent?

Reply:
I'm not convinced this variant is necessary. Using Constant and matching on the value should be enough information.

Reply:
Inline constants, like other constants, can be lowered to actual patterns in THIR. We could keep them as constants here and lower to patterns in both MIR building and exhaustiveness checking, but I don't think that's any better.

Reply:
Previously we discarded the original constant entirely and keep only the Pat (unless the constant can't be converted). Here you're keeping the original constant around for unsafety checking if I follow correctly? Could this check be done beforehand instead? Otherwise, could you add a comment explaining why you need to keep the constant around?

Reply:
I guess my question is: is there a reason that the pattern-handling code should care about the difference between inline and normal consts? Evaluation order maybe?